doc/man/man3/Shader.3
.TH "Shader" 3 "Wed Mar 6 2019" "Version 1.0" "Epsilon Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Shader \- 
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBShader\fP (const char *, const char *)"
.br
.ti -1c
.RI "\fBShader\fP (const char *, const char *, const char *)"
.br
.ti -1c
.RI "\fBShader\fP (const char *, const char *, const char *, const char *, const char *)"
.br
.ti -1c
.RI "\fBShader\fP (const char *)"
.br
.ti -1c
.RI "void \fBUse\fP (void)"
.br
.ti -1c
.RI "void \fBFree\fP (void)"
.br
.ti -1c
.RI "GLuint \fBgenerateFragmentProgram\fP (std::string)"
.br
.ti -1c
.RI "GLuint \fBgenerateVertexProgram\fP (std::string)"
.br
.ti -1c
.RI "GLuint \fBgenerateGeometryProgram\fP (std::string)"
.br
.ti -1c
.RI "GLuint \fBgetProgramID\fP ()"
.br
.ti -1c
.RI "std::string \fBgetPath\fP ()"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, int data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, float data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, glm::vec2 data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, glm::vec3 data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, glm::vec4 data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, glm::mat3 data)"
.br
.ti -1c
.RI "void \fBPushUniform\fP (std::string name, glm::mat4 data)"
.br
.in -1c
.SS "Data Fields"

.in +1c
.ti -1c
.RI "GLuint \fBMVP_Location\fP"
.br
.ti -1c
.RI "GLuint \fBWorldTransform_Location\fP"
.br
.ti -1c
.RI "GLuint \fBProjection_Location\fP"
.br
.ti -1c
.RI "GLuint \fBView_Location\fP"
.br
.ti -1c
.RI "GLuint \fBViewDirection_Location\fP"
.br
.ti -1c
.RI "GLuint \fBViewPosition_Location\fP"
.br
.ti -1c
.RI "GLuint \fBLightSpaceMatrix_Location\fP"
.br
.ti -1c
.RI "GLuint \fBModelView3x3Matrix_Location\fP"
.br
.ti -1c
.RI "GLuint \fBNormalMatrix_Location\fP"
.br
.ti -1c
.RI "GLuint \fBLightDirection_Location\fP"
.br
.ti -1c
.RI "GLuint \fBClipPlane_Location\fP"
.br
.ti -1c
.RI "GLuint \fBTime_Location\fP"
.br
.ti -1c
.RI "GLuint \fBtexture_diffuse_Location\fP"
.br
.ti -1c
.RI "GLuint \fBtexture_specular_Location\fP"
.br
.ti -1c
.RI "GLuint \fBtexture_normal_Location\fP"
.br
.ti -1c
.RI "GLuint \fBtexture_height_Location\fP"
.br
.ti -1c
.RI "GLuint \fBskybox_Location\fP"
.br
.ti -1c
.RI "GLuint \fBviewPos_Location\fP"
.br
.ti -1c
.RI "GLuint \fBPrevViewPos_Location\fP"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBgetUniformsLocations\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "GLuint \fBProgramID\fP"
.br
.ti -1c
.RI "std::string \fBPath\fP"
.br
.ti -1c
.RI "std::string \fBvPath\fP"
.br
.ti -1c
.RI "std::string \fBfPath\fP"
.br
.ti -1c
.RI "std::unordered_map< std::string, GLuint > \fBm_Attributes\fP"
.br
.ti -1c
.RI "std::unordered_map< std::string, GLuint > \fBm_Uniforms\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
========= Copyright Survtech, All rights reserved\&. ============//
.PP
Purpose: 
.PP
 
.PP
Definition at line 19 of file Shader\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Shader::Shader (const char * vertex, const char * fragment)"
========= Copyright Survtech, All rights reserved\&. ============//
.PP
Purpose: 
.PP
 Link the program
.PP
Check the program 
.PP
Definition at line 17 of file Shader\&.cpp\&.
.SS "Shader::Shader (const char * vertex, const char * fragment, const char * geometry)"
Link the program
.PP
Check the program 
.PP
Definition at line 60 of file Shader\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "GLuint Shader::generateFragmentProgram (std::string pShaderString)"
Compile Fragment \fBShader\fP
.PP
Check Fragment \fBShader\fP 
.PP
Definition at line 105 of file Shader\&.cpp\&.
.SS "GLuint Shader::generateGeometryProgram (std::string pShaderString)"
Check \fBVertex\fP \fBShader\fP 
.PP
Definition at line 152 of file Shader\&.cpp\&.
.SS "GLuint Shader::generateVertexProgram (std::string pShaderString)"
Check \fBVertex\fP \fBShader\fP 
.PP
Definition at line 129 of file Shader\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Epsilon Engine from the source code\&.


src/Shader.cpp
///========= Copyright Survtech, All rights reserved. ============//
///
/// Purpose:
///
///=============================================================================
#include <Shader.h>
#include <stdio.h>
#include <string>
#include <vector>
#include <iostream>
#include <fstream>
#include <algorithm>
#include <Log.h>
#include <Includes.h>
#include <OpenGL/ShaderPreProcessor.h>

Shader::Shader(const char* vertex, const char* fragment)
{
    Path = std::string(vertex);
    vPath = vertex;
    fPath = fragment;

	std::string VertexShaderCode = GLSLPreProcessor::PreProcessGLSL(std::string(vertex));

	std::string FragmentShaderCode = GLSLPreProcessor::PreProcessGLSL(std::string(fragment));

	GLuint VertexShaderID = generateVertexProgram(VertexShaderCode);

	GLuint FragmentShaderID = generateFragmentProgram(FragmentShaderCode);


		std::cout << Path << std::endl;

	/// Link the program
    ProgramID = glCreateProgram();
	glAttachShader(ProgramID, VertexShaderID);
	glAttachShader(ProgramID, FragmentShaderID);
	glLinkProgram(ProgramID);

	GLint Result = GL_FALSE;
	int InfoLogLength;
	/// Check the program
	glGetProgramiv(ProgramID, GL_LINK_STATUS, &Result);
	glGetProgramiv(ProgramID, GL_INFO_LOG_LENGTH, &InfoLogLength);
	if ( InfoLogLength > 0 ){
		char ProgramErrorMessage[InfoLogLength+1];
		glGetProgramInfoLog(ProgramID, InfoLogLength, NULL, ProgramErrorMessage);
		//printf("%s\n", &ProgramErrorMessage[0]);
		std::cout << ProgramErrorMessage << std::endl;
		Global::Log::WriteToLog("Error while compiling: " + Helpers::removeExtension(Path));
		Global::Log::WriteToLog(std::string(ProgramErrorMessage));
	}
	glDeleteShader(VertexShaderID);
	glDeleteShader(FragmentShaderID);
	this->getUniformsLocations();

}

Shader::Shader(const char* vertex, const char* fragment, const char* geometry)
{
    Path = std::string(vertex);
    vPath = vertex;
    fPath = fragment;

	std::string VertexShaderCode = GLSLPreProcessor::PreProcessGLSL(std::string(vertex));

	std::string FragmentShaderCode = GLSLPreProcessor::PreProcessGLSL(std::string(fragment));

	std::string GeometryShaderCode = GLSLPreProcessor::PreProcessGLSL(std::string(geometry));

	GLuint VertexShaderID = generateVertexProgram(VertexShaderCode);

	GLuint FragmentShaderID = generateFragmentProgram(FragmentShaderCode);

	GLuint GeometryShaderID = generateGeometryProgram(GeometryShaderCode);

	/// Link the program
	//printf("Linking program\n");
    ProgramID = glCreateProgram();
	glAttachShader(ProgramID, VertexShaderID);
	glAttachShader(ProgramID, FragmentShaderID);
	glAttachShader(ProgramID, GeometryShaderID);
	glLinkProgram(ProgramID);

	GLint Result = GL_FALSE;
	int InfoLogLength;
	/// Check the program
	glGetProgramiv(ProgramID, GL_LINK_STATUS, &Result);
	glGetProgramiv(ProgramID, GL_INFO_LOG_LENGTH, &InfoLogLength);
	if ( InfoLogLength > 0 ){
		char ProgramErrorMessage[InfoLogLength+1];
		glGetProgramInfoLog(ProgramID, InfoLogLength, NULL, ProgramErrorMessage);
		//printf("%s\n", &ProgramErrorMessage[0]);
		std::cout << ProgramErrorMessage << std::endl;
		Global::Log::WriteToLog("Error while compiling: " + Helpers::removeExtension(Path));
		Global::Log::WriteToLog(std::string(ProgramErrorMessage));
	}
	glDeleteShader(VertexShaderID);
	glDeleteShader(FragmentShaderID);
	glDeleteShader(GeometryShaderID);
	this->getUniformsLocations();
}

GLuint Shader::generateFragmentProgram(std::string pShaderString)
{
	GLint Result = GL_FALSE;
	int InfoLogLength;
    /// Compile Fragment Shader
	//printf("Compiling shader : %s\n", fragment);
	GLuint FragmentShaderID = glCreateShader(GL_FRAGMENT_SHADER);
	char const * FragmentSourcePointer = pShaderString.c_str();
	glShaderSource(FragmentShaderID, 1, &FragmentSourcePointer , NULL);
	glCompileShader(FragmentShaderID);

	/// Check Fragment Shader
	glGetShaderiv(FragmentShaderID, GL_COMPILE_STATUS, &Result);
	glGetShaderiv(FragmentShaderID, GL_INFO_LOG_LENGTH, &InfoLogLength);
	if ( InfoLogLength > 0 ){
		std::vector<char> FragmentShaderErrorMessage(InfoLogLength+1);
		glGetShaderInfoLog(FragmentShaderID, InfoLogLength, NULL, &FragmentShaderErrorMessage[0]);
		printf("%s\n", &FragmentShaderErrorMessage[0]);
	}

    Global::Log::WriteToLog("Fragment shader: " + std::string(fPath) + " was successfully compiled.");
	return FragmentShaderID;
}

GLuint Shader::generateVertexProgram(std::string pShaderString)
{
	GLint Result = GL_FALSE;
	int InfoLogLength;
	GLuint VertexShaderID = glCreateShader(GL_VERTEX_SHADER);
	char const * VertexSourcePointer = pShaderString.c_str();
	glShaderSource(VertexShaderID, 1, &VertexSourcePointer , NULL);
	glCompileShader(VertexShaderID);

	/// Check Vertex Shader
	glGetShaderiv(VertexShaderID, GL_COMPILE_STATUS, &Result);
	glGetShaderiv(VertexShaderID, GL_INFO_LOG_LENGTH, &InfoLogLength);
	if ( InfoLogLength > 0 ){
		std::vector<char> VertexShaderErrorMessage(InfoLogLength+1);
		glGetShaderInfoLog(VertexShaderID, InfoLogLength, NULL, &VertexShaderErrorMessage[0]);
		printf("%s\n", &VertexShaderErrorMessage[0]);
	}

    Global::Log::WriteToLog("Vertex shader: " + std::string(vPath) + " was successfully compiled.");
	return VertexShaderID;
}


GLuint Shader::generateGeometryProgram(std::string pShaderString)
{
	GLint Result = GL_FALSE;
	int InfoLogLength;
	GLuint GeometryShaderID = glCreateShader(GL_GEOMETRY_SHADER);
	char const * GeometrySourcePointer = pShaderString.c_str();
	glShaderSource(GeometryShaderID, 1, &GeometrySourcePointer , NULL);
	glCompileShader(GeometryShaderID);

	/// Check Vertex Shader
	glGetShaderiv(GeometryShaderID, GL_COMPILE_STATUS, &Result);
	glGetShaderiv(GeometryShaderID, GL_INFO_LOG_LENGTH, &InfoLogLength);
	if ( InfoLogLength > 0 ){
		std::vector<char> GeometryShaderErrorMessage(InfoLogLength+1);
		glGetShaderInfoLog(GeometryShaderID, InfoLogLength, NULL, &GeometryShaderErrorMessage[0]);
		printf("%s\n", &GeometryShaderErrorMessage[0]);
	}

	return GeometryShaderID;
}
void Shader::getUniformsLocations()
{

    // NOTE(Joey): query the number of active uniforms and attributes
        int nrAttributes, nrUniforms;
        glGetProgramiv(ProgramID, GL_ACTIVE_ATTRIBUTES, &nrAttributes);
        glGetProgramiv(ProgramID, GL_ACTIVE_UNIFORMS, &nrUniforms);
        //m_Attributes.resize(nrAttributes);
        //m_Uniforms.resize(nrUniforms);

        // NOTE(Joey): iterate over all active attributes
        char buffer[128];
        int Size;
        for (unsigned int i = 0; i < nrAttributes; ++i)
        {
            GLenum glType;
            glGetActiveAttrib(ProgramID, i, sizeof(buffer), 0, &Size, &glType, buffer);

            this->m_Attributes[std::string(buffer)] = glGetAttribLocation(ProgramID, buffer);
            //std::cout << "Attribute: " << i << " :: " << buffer << std::endl;
        }

        // NOTE(Joey): iterate over all active uniforms
        for (unsigned int i = 0; i < nrUniforms; ++i)
        {
            GLenum glType;
            glGetActiveUniform(ProgramID, i, sizeof(buffer), 0, &Size, &glType, buffer);

            this->m_Uniforms[std::string(buffer)] = glGetUniformLocation(ProgramID, buffer);
            //std::cout << "Uniform: " << i << " :: " << buffer << std::endl;
        }

    this->MVP_Location = glGetUniformLocation(this->ProgramID, "MVP");
    this->WorldTransform_Location = glGetUniformLocation(this->ProgramID, "model");
    this->Projection_Location = glGetUniformLocation(this->ProgramID, "projection");
    this->View_Location = glGetUniformLocation(this->ProgramID, "view");
    this->LightSpaceMatrix_Location = glGetUniformLocation(this->ProgramID, "lightSpaceMatrix");
    this->viewPos_Location = glGetUniformLocation(this->ProgramID, "viewPos");
    this->PrevViewPos_Location = glGetUniformLocation(this->ProgramID, "PrevView");
    //this->MVP_Location = glGetUniformLocation(this->ProgramID, "MVP");
    //this->MVP_Location = glGetUniformLocation(this->ProgramID, "MVP");
}
